src/Avalonia.Controls/WrapPanel.cs

```diff
@@ -176,40 +208,33 @@ protected override Size MeasureOverride(Size constraint)
                 itemWidthSet ? itemWidth : constraint.Width,
                 itemHeightSet ? itemHeight : constraint.Height);
 
-            for (int i = 0, count = children.Count; i < count; i++)
+            for (int i = 0, count = children.Count; i < count; ++i)
             {
                 var child = children[i];
                 // Flow passes its own constraint to children
                 child.Measure(childConstraint);
 
                 // This is the size of the child in UV space
-                var sz = new UVSize(orientation,
+                UVSize childSize = new UVSize(orientation,
                     itemWidthSet ? itemWidth : child.DesiredSize.Width,
                     itemHeightSet ? itemHeight : child.DesiredSize.Height);
 
-                if (MathUtilities.GreaterThan(curLineSize.U + sz.U, uvConstraint.U)) // Need to switch to another line
+                if (MathUtilities.GreaterThan(curLineSize.U + childSize.U + (curLineSize.U == 0 ? 0 : itemSpacing), uvConstraint.U)) // Need to switch to another line
```

Comment:
I'm leaving this comment on this line, but this affects several other paths in `Measure/Arrange`:
The spacing shouldn't be added for items having `IsVisible=false`, matching `StackPanel`.